rust/connlib/tunnel/src/client.rs

```diff
@@ -45,18 +45,22 @@ const DNS_SENTINELS_V6: &str = "fd00:2021:1111:8000:100:100:111:0/120";
 // is 30 seconds. See resolvconf(5) timeout.
 const IDS_EXPIRE: std::time::Duration = std::time::Duration::from_secs(60);
 
-impl<CB> ClientTunnel<CB>
-where
-    CB: Callbacks + 'static,
-{
+impl ClientTunnel {
     pub fn set_resources(&mut self, resources: Vec<ResourceDescription>) {
         self.role_state.set_resources(resources);
-        self.callbacks.on_update_routes(
-            self.role_state.routes().filter_map(utils::ipv4).collect(),
-            self.role_state.routes().filter_map(utils::ipv6).collect(),
-        );
-        self.callbacks
-            .on_update_resources(self.role_state.resources());
+
+        // FIXME: It would be good to add this event from _within_ `set_resources` but we don't want to emit duplicates.
```

Comment:
```suggestion
        // FIXME: It would be good to add this event from _within_ `ClientState` but we don't want to emit duplicates.
```

Reply:
isn't this the same if you move the below 2 functions to the end of `set_routes`? We needed this outside just because we needed access to `callbacks`.

Reply:
I guess the problem is if `add_resources` and `remove_resources` emit the same event this would emit duplicates. Right? One solution would be to implement `add_resources` and `remove_resources` based on `set_resources` instead of the other way around.

Reply:
> I guess the problem is if `add_resources` and `remove_resources` emit the same event this would emit duplicates. Right? One solution would be to implement `add_resources` and `remove_resources` based on `set_resources` instead of the other way around.

Yes! I realised we can do #5886 too, maybe that makes things a bit easier.